### DISCUSSION TITLE
Is it possible to deploy a Django application that has no frontend component (i.e. is just a backend API)?

### DISCUSSION THREAD

**Discussion by sherrellbc** [vercel/vercel#11081] [Help] [answered]
### Question

I have an app that uses Next.js in the frontend and a Django backend. I was able to get the frontend deployed simply enough but I am not sure how to deploy the backend.

There is quite a lot of _magic_ happening when using the Vercel deploy so I am not sure how to debug this. According to the Vercel CLI the deploy has completed successfully (I just used the no-arg `vercel` tool directly from the Django project root directory). The deploy tool appears to have figured out the project was based on Django/python and installed the missing dependencies and such. But now what? 

- When you run a Django server it listens on port 8000 by default. Does Vercel set up a reverse proxy that forwards :80 and :443 connections on to the right place?
- As far as I am aware, Django does not itself natively support SSL connections (but the default deploy sets up certs and uses it). I suppose this is managed somehow?

The behavior I see is that the deployment appears to complete successfully but when I try to hit the domain name in a browser I just see the NOT_FOUND error page:
```
404: NOT_FOUND
Code: NOT_FOUND
ID: sfo1::vvzfd-1705989958921-ecd6b9b09720
```

But this is somewhat expected since the Django project is just a backend and does not actually render HTML? I did find a simple [Django template example](https://vercel.com/templates/python/django-hello-world) but it's also not overly helpful. It walks you through configuring the Django app in a very ordinary (non-Vercel oriented) way, then goes on to say you should just deploy the app and it magically works.

My intent is to have my frontend deployed separately (i.e. two Vercel projects). The frontend will and environment vars that indicate the URL of its backend. Is there a problem with this approach?

I am thoroughly confused on what to do next.

**Comment by leerob** [comment, marked as answer]
https://vercel.com/templates/python/django-hello-world

**Reply by sherrellbc** [reply]
You can see I linked this same repo in my original question. The difference being that I was using this strictly as an _API_, whereas the example uses it as a frontend/backend server (notably a difference in what port the server should listen on). And either way, this page does not efficiently explain the steps for proper deploy. It explains how to configure an example Django project and run it locally. Then it ends with "now deploy it to Vercel". 

But how does that work? When run locally the Django server listens by default on port 8000 on _localhost_. But indeed, somehow if you deploy your Python project properly all of this _just works_ magically with no real explanation. My expectation of deploying my Django project would be that it also listens on port 8000 on the Vercel host. This is not actually what happens. Does Vercel implicitly add some form of reverse proxy? It's unclear. Somehow requests to port 443 make it to the Django server. But none of this is explained.

Eventually I was able to get it to work, but only after finding the [Python Runtime](https://vercel.com/docs/functions/serverless-functions/runtimes/python) page. Specifically the [Advanced Python Usage](https://vercel.com/docs/functions/serverless-functions/runtimes/python#advanced-python-usage) section which has:

> Advanced usage of the Python Runtime, such as with Flask and Django, requires some configuration.

and 

> The entry point of this Runtime is a glob matching .py source files with one of the following variables defined:
>
> handler that inherits from the BaseHTTPRequestHandler class
> app that exposes a WSGI or ASGI Application

So I went to the [Django source example](https://github.com/vercel/examples/tree/main/python/django) and found [this small change to the default Django project](https://github.com/vercel/examples/blob/main/python/django/vercel_app/wsgi.py#L16) that satisfies the second option. But then it also has a [vercel.json](https://github.com/vercel/examples/blob/main/python/django/vercel.json) file. Is this required? I don't know. Because it's not documented.

Anyway, after editing my local `wsgi.py` file and including the `vercel.json` the application was properly deployed. I'm not sure if the latter is required.
